File: ios/NMOSevaLedger/APIClient.swift
import Foundation

enum APIError: Error { case invalidURL, badResponse }

struct APIClient {
    static func upload(events: [LedgerEvent], settings: SettingsStore) async throws {
        guard !events.isEmpty else { return }
        guard let url = URL(string: settings.backendURL) else { throw APIError.invalidURL }

        let eventObjects: [[String: Any?]] = events.map { event in
            [
                "eventId": event.eventId,
                "amount": event.amount,
                "donorName": event.donorName,
                "expenseNote": event.expenseNote,
                "senderHint": event.senderHint,
                "transactionRef": event.transactionRef,
                "receivedAt": event.receivedAt,
                "sourceApp": event.sourceApp,
                "direction": event.direction.rawValue,
                "donationStatus": event.donationStatus.rawValue,
                "expenseStatus": event.expenseStatus.rawValue,
                "reconciled": event.reconciled
            ]
        }

        let cleanEvents = eventObjects.map { dict in
            dict.reduce(into: [String: Any]()) { result, pair in
                result[pair.key] = pair.value ?? NSNull()
            }
        }
        let body: [String: Any] = [
            "collectorCode": settings.collectorCode.uppercased(),
            "token": settings.collectorToken,
            "events": cleanEvents
        ]

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else { throw APIError.badResponse }
        let json = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
        guard json?["ok"] as? Bool == true else { throw APIError.badResponse }
    }
}

File: app/src/main/java/org/nmo/seva/collector/ApiClient.kt
package org.nmo.seva.collector

import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

object ApiClient {
    fun upload(url: String, collectorCode: String, token: String, payments: List<Payment>): Boolean {
        if (payments.isEmpty()) return true
        val events = JSONArray()
        payments.forEach { p ->
            events.put(JSONObject().apply {
                put("eventId", p.eventId)
                put("amount", p.amount)
                put("donorName", p.donorName ?: JSONObject.NULL)
                put("expenseNote", p.expenseNote ?: JSONObject.NULL)
                put("senderHint", p.senderHint ?: JSONObject.NULL)
                put("transactionRef", p.transactionRef ?: JSONObject.NULL)
                put("receivedAt", p.receivedAt)
                put("sourceApp", p.sourceApp)
                put("direction", p.direction.name)
                put("donationStatus", p.donationStatus.name)
                put("expenseStatus", p.expenseStatus.name)
                put("reconciled", p.reconciled)
            })
        }
        val body = JSONObject().apply {
            put("collectorCode", collectorCode)
            put("token", token)
            put("events", events)
        }.toString()

        val conn = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = 15_000
            readTimeout = 15_000
            doOutput = true
            setRequestProperty("Content-Type", "application/json; charset=utf-8")
        }
        return try {
            conn.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
            val code = conn.responseCode
            val stream = if (code in 200..299) conn.inputStream else conn.errorStream
            val response = stream?.bufferedReader()?.use { it.readText() }.orEmpty()
            code in 200..299 && runCatching { JSONObject(response).optBoolean("ok", false) }.getOrDefault(false)
        } finally {
            conn.disconnect()
        }
    }
}
